assets/app/scripts/services/labelFilter.js

```diff
@@ -0,0 +1,337 @@
+'use strict';
+
+angular.module('openshiftConsole')
+.factory('LabelFilter', [function() {
+  function LabelFilter() {
+    this._existingLabels = {};
+    this._activeFilters = {};
+    this._onActiveFiltersChangedCallbacks = $.Callbacks();
+  }
+
+  LabelFilter.prototype.createLabelSuggestionsFromResources = function(items, map) {
+    // check if we are extracting from a single item or a hash of items
+    if (items.id || (items.metadata && items.metadata.name)) {
+      this._extractLabelsFromItem(items, map);
+    }
+    else {
+      var self = this;
+      $each(items, function(key, item) {
+        self._extractLabelsFromItem(item, map);
+      });
+    }
+  };
+
+  LabelFilter.prototype.setLabelSuggestions = function(suggestions) {
+    this._existingLabels = suggestions;
+  };
+
+  LabelFilter.prototype._extractLabelsFromItem = function(item, map) {
+    // TODO temporary until we handle everything with item.metadata.labels
+    var labels = item.labels;
+    if (item.metadata) {
+      labels = item.metadata.labels;
+    }
+
+    var self = this;
+    $each(labels, function(key, value) {
+      if (!map[key]) {
+        map[key] = [];
+      }
+      map[key].push({value: value});
+    });
+  };
+
+  LabelFilter.prototype.onActiveFiltersChanged = function(callback) {
+    this._onActiveFiltersChangedCallbacks.add(callback);
+  };
+
+  LabelFilter.prototype.filterResources = function(resources) {
+    var filteredResources = {};
+    var self = this;
+    $each(resources, function(resId, resource) {
+      if (self.isResourceIncludedInActiveFilters(resource)) {
+        filteredResources[resId] = resource;
+      }
+    });
+    return filteredResources;
+  };
+
+  LabelFilter.prototype.isResourceIncludedInActiveFilters = function(resource) {
+    if (!resource) {
+      return false;
+    }
+    var labels = resource.labels || {};
+    if (resource.metadata) {
+      labels = resource.metadata.labels || {};
+    }
+    for (var id in this._activeFilters) {
```

Comment:
Separate out the object that represents a label selector, and the logic to decide whether that matches a set of labels. Shouldn't be tied so closely to a LabelFilter widget (we'll want to be able to parse a label selector from an API object later). A LabelFilter widget should be able to return a LabelSelector object representing its active filters. Then `LabelSelector.select(resources)` can select resources (probably calling `LabelSelector.matches(resource)` internally)